---
title: "Assignment analysis of mixture samples"
author: "Thierry Gosselin"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

Next-generation sequencing techniques that reduce the size of the genome (e.g. genotype-by-sequencing (GBS) and
restriction-site-associated DNA sequencing (RADseq)) produce huge numbers of markers that hold great potential and promises for assignment analysis. After hitting the bioinformatic wall with the different workflows you'll likely end up with several folders containing whitelist and blacklist of markers and individuals, data sets with various *de novo* and/or filtering parameters and missing data. This reality of GBS/RAD data is quite hard on GUI software traditionally used for assignment analysis. The end results is usually poor data exploration, constrained by time, and poor reproducibility.

**assigner** was tailored to make it easy to conduct assignment analysis using GBS/RAD data within R. Additionally, combining the use of tools like [RStudio](https://www.rstudio.com) and [GitHub](https://github.com) will make effortless documenting your workflows and pipelines.


## assignment_mixture

This document introduces you to the basic set of tools of `assignment_mixture` 
function, and shows you how to apply them to conduct assignment analysis with 
baseline and mixture samples. The argument are presented in the order they are 
used in the function.

**1. data** : 4 types of input files are available. The input file should contain
both *baseline* and *mixture* samples.
    
* VCF file format (Danecek et al. 2011) produced by [STACKS](http://catchenlab.life.illinois.edu/stacks/) (Catchen et al. 2011, 2013). e.g. *batch_1.vcf*. For huge VCF files, the use of 
the PLINK `tped` format is faster and recommended. 
For this, use [VCFTOOLS](http://vcftools.sourceforge.net/) with `--plink-tped`
option to convert the large VCF file.

* haplotypes data frame file produced by [STACKS](http://catchenlab.life.illinois.edu/stacks/) (Catchen et al. 2011, 2013). e.g. *batch_x.haplotypes.tsv*.

* PLINK tped/tfam format (Purcell et al. 2007) e.g. *data.assignment.tped*.

    This format is useful for large dataset. The first 2 columns of the `tfam` 
    file will be used for the `strata` argument (explained below), 
    unless a new `strata` argument is provided. Columns 1, 3 and 4 
    of the `tped` are discarded. The remaining columns correspond to the 
    genotypes in the format `01/04`, where `A = 01, C = 02, G = 03 and T = 04`. 
    For `A/T` format, use [PLINK](http://pngu.mgh.harvard.edu/~purcell/plink/)
    or bash to convert. For `.ped` file conversion 
    to `.tped` use [PLINK](http://pngu.mgh.harvard.edu/~purcell/plink/)
    with `--recode transpose` option. 
    
    For the function argument, you only provide the name of the `tped` file, 
    the function will take care of finding and importing the `tfam`. Both files
    must have the same name, e.g. `data = "tuna.tped"` then the `tfam` file must
    be named: `tuna.tfam`.

* data frame of genotypes with tab separated columns. 
    
    The 1st column is the `POP_ID`, 2nd column the `INDIVIDUALS`. The same naming
    is required. The first 2 columns of the `tfam` 
    file will be used for the `strata` argument (explained below), 
    unless a new `strata` argument is provided.
    The remaining columns are the markers IDs (naming not important),
    these columns contains the genotypes in 3 digits format per allele with no 
    space between the 2 alleles (e.g. 235240 : allele1 = 235 and allele2 = 240).
    Missing genotypes are coded `0` or `000000`. Note: the `POP_ID` column 
    can be any hierarchical grouping.


POP_ID |INDIVIDUALS|LOCUS_1|LOCUS_2|LOCUS_3
:-------:|:-------:|:-------:|:-------:|:-------:|
BAJ|YFT-BAJ-ADU-2016-001|001002|003003|001004
BAJ|YFT-BAJ-ADU-2016-002|002002|001003|002004
IND|YFT-IND-ADU-2016-003|001001|002003|001003
IND|YFT-IND-ADU-2016-004|003004|004004|002004

The argument and options:
```{r}
data = "batch_1.vcf" # or
data = "batch_1.haplotypes.tsv"
data = "tuna.tped" 
data = "tuna.tsv"
```

**2. mixture** 
The mixture file is a standard text file with 1 column containing the mixture 
individual IDs. The column name is `INDIVIDUALS` and the remaining rows are the IDs.

Make sure the lines endings of your files is in the UNIX (LF) format and not Classic Mac (CR) or 
Windows (CRLF). Use [flip](https://ccrma.stanford.edu/~craig/utility/flip/) to convert files format.

```{r}
mixture = "mixture.tuna.txt"
```

**3. whitelist.markers** 
Once the dataset is imported in R, the function will look to filter the markers 
if the argument `whitelist.markers` is provided. Useful if your have filtered your
dataset downstream of the software tha generated the dataset. The whitelist can contain up 
to 3 columns: `CHROM`, `LOCUS` and `POS`. for those working with VCF, 
the column `ID` in the VCF is the `LOCUS` identification. The `POS` column 
represent the position of the SNP. Below an example with 3 columns:

CHROM |LOCUS|POS
:-------:|:-------:|:-------:|
1|255|34
1|255|72
1|13290|63

```{r}
whitelist.markers = "mixture.tuna.txt"
```


**4. blacklist.id** 
If you need to remove individuals from the dataset before conducting the 
assignment analysis use the argument `blacklist.id`. The file must contain a 
single column with the header `INDIVIDUALS`.

```{r}
blacklist.id = "blacklisted.individuals.txt"
```

**5. strata, pop.id.start, pop.id.end, pop.levels and pop.labels** 
These argument exist to make your life easier with individuals, sampling sites,
populations and other stratification or grouping you might want to use, without 
messing with your input file. There are severals ways you can use those arguments
to identify your populations or groupings.

* `stata`: For PLINK and data frame of genotypes file type, the grouping of individuals is already 
present inside the file (not with VCF or haplotype file). 
However, if you decide to modify the 
groupings (e.g. after having conducted a DAPC analysis or a K-mean clustering), 
or make the VCF and haplotype file population-wise, the `strata` argument 
will be usefull. This argument points to a tab delimited 
file with 2 columns containing the headers: `INDIVIDUALS` and `STRATA`. 
If a strata file is specified, the strata file will have precedence over the 
`POP_ID` of the data frame of genotypes or the PLINK file. 
The `STRATA` column can be any hierarchical grouping you want to use for the 
assignment analysis.

* `pop.id.start` and `pop.id.end`:
If your individuals have the grouping inside their names, you can use the 
`pop.id.start` and `pop.id.end` arguments to isolate the grouping. 
e.g. if your individuals are identified in this form : 
`SPECIES-POPULATION-MATURITY-YEAR-ID = CHI-QUE-ADU-2014-020`, 
`pop.id.start = 5` and `pop.id.end = 7`. If you prefer the `strata` argument, 
just don't use these 2 arguments.

* `pop.levels`
Argument that correspond to a character string of your groupings.
This argument is required to order the populations or groupings inside the table,
so that you don't have to do it manually. 
e.g. you have 3 populations: HAY, ANG, and FEG, normally R would put them in alphabetic order,
however you want them in this order, corresponding to upstream -> downstream 
locations:
    ```{r}
pop.levels = c("HAY", "ANG", "FEG")
```

* `pop.labels`
The `pop.labels` argument is optional and enables you to change the name of the 
`pop.levels` grouping identified earlier. 
Also useful if you want to combined sites or populations.
    ```{r}
# to combined HAY and ANG into a new grouping: HAY-ANG
pop.levels = c("HAY", "ANG", "FEG")
pop.labels = c("HAY-ANG", "HAY-ANG", "FEG")
```

## Styles

The `html_vignette` template includes a basic CSS theme. To override this theme you can specify your own CSS in the document metadata as follows:

    output: 
      rmarkdown::html_vignette:
        css: mystyles.css

## Figures

The figure sizes have been customised so that you can easily put two images side-by-side. 

```{r, fig.show='hold'}
plot(1:10)
plot(10:1)
```

You can enable figure captions by `fig_caption: yes` in YAML:

    output:
      rmarkdown::html_vignette:
        fig_caption: yes

Then you can use the chunk option `fig.cap = "Your figure caption."` in **knitr**.

## More Examples

You can write math expressions, e.g. $Y = X\beta + \epsilon$, footnotes^[A footnote here.], and tables, e.g. using `knitr::kable()`.

```{r, echo=FALSE, results='asis'}
knitr::kable(head(mtcars, 10))
```

Also a quote using `>`:

> "He who gives up [code] safety for [code] speed deserves neither."
([via](https://twitter.com/hadleywickham/status/504368538874703872))
